src/mbgl/style/collection.hpp

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include <mbgl/util/immutable.hpp>
+#include <mbgl/util/optional.hpp>
+
+#include <memory>
+#include <string>
+
+namespace mbgl {
+namespace style {
+
+/*
+    Manages an ordered collection of elements and their `Immutable<Impl>`s. The latter is
+    itself stored in an Immutable container. Using immutability at the collection level
+    allows us to short-circuit significant portions of the RenderStyle update logic via
+    a simple pointer equality check, greatly improving performance.
+
+    Element types are required to have:
+
+      * An `Impl` inner class type
+      * An `Immutable<Impl> baseImpl` member
+      * A `std::string getID() const` method
+*/
+template <class T>
+class Collection {
+public:
+    using Impl = typename T::Impl;
+    using WrapperVector = std::vector<std::unique_ptr<T>>;
+    using ImmutableVector = Immutable<std::vector<Immutable<Impl>>>;
+
+    Collection();
+
+    std::size_t size() const;
+    std::size_t index(const std::string&) const;
+    T* get(const std::string&) const;
+
+    std::vector<T*> getWrappers() const;
+    ImmutableVector getImpls() const { return impls; }
+
+    auto begin() const { return wrappers.begin(); }
+    auto end() const { return wrappers.end(); }
+
+    void clear();
+
+    T* add(std::unique_ptr<T>, const optional<std::string>& = {});
+    std::unique_ptr<T> remove(const std::string&);
+
+    // Must be called whenever an element of the collection is internally mutated.
+    // Typically, each element permits registration of an observer, and the observer
+    // should call this method.
+    void update(const T&);
+
+private:
+    WrapperVector wrappers;
+    ImmutableVector impls;
+};
+
+template <class T>
+Collection<T>::Collection()
+    : impls(makeMutable<std::vector<Immutable<Impl>>>()) {
+}
+
+template <class T>
+std::size_t Collection<T>::size() const {
+    return wrappers.size();
+}
+
+template <class T>
+std::size_t Collection<T>::index(const std::string& id) const {
+    return std::find_if(wrappers.begin(), wrappers.end(), [&](const auto& e) {
+        return e->getID() == id;
+    }) - wrappers.begin();
```

Comment:
It returns `end - begin`, i.e. the collection size. This is checked by callers where needed. The only callers are in `Collection` itself so I'll make it private.